include/seqan/index/index_bifm_stree.h

```diff
@@ -0,0 +1,483 @@
+    }
+
+//____________________________________________________________________________
+
+    template <typename TSpec2>
+    inline Iter const &
+    operator = (Iter<TBiIndex, VSTree<TopDown<TSpec2> > > const &_origin)
+    {
+        fwdIter = *(&_origin.fwdIter);
+        bwdIter = *(&_origin.bwdIter);
+        return *this;
+    }
+};
+
+// TopDown-Iterator with history stack for bidirectional FM index
+template <typename TText, class TOccSpec, typename TLengthSum, typename TBidirectional, typename TSpec>
+class Iter<Index<TText, BidirectionalFMIndex<TOccSpec, FMIndexConfig<TOccSpec, TLengthSum, TBidirectional> > >, VSTree<TopDown<ParentLinks<TSpec> > > >
+{
+public:
+    typedef Index<TText, BidirectionalFMIndex<TOccSpec, FMIndexConfig<TOccSpec, TLengthSum, TBidirectional> > >  TBiIndex;
+
+    typedef typename BiFMReversedText<TText>::Type                                                            TRevText;
+
+    typedef Index<TText, FMIndex<TOccSpec, FMIndexConfig<TOccSpec, TLengthSum, FMBidirectional> > >           TFwdIndex;
+    typedef Index<TRevText, FMIndex<TOccSpec, FMIndexConfig<TOccSpec, TLengthSum, FMBidirectional> > >        TRevIndex;
+
+    typedef Iter<TFwdIndex, VSTree<TopDown<ParentLinks<TSpec> > > >                                           TFwdIndexIter;
+    typedef Iter<TRevIndex, VSTree<TopDown<ParentLinks<TSpec> > > >                                           TRevIndexIter;
+
+    TFwdIndexIter    fwdIter;
+    TRevIndexIter    bwdIter;
+
+//____________________________________________________________________________
+
+    Iter():
+        fwdIter(),
+        bwdIter() {}
+
+    Iter(TBiIndex &_index):
+        fwdIter(*(&_index.fwd)),
+        bwdIter(*(&_index.rev))
+    {
+        fwdIter.setRevIter(bwdIter);
+        bwdIter.setRevIter(fwdIter);
+    }
+
+    Iter(TBiIndex &_index, MinimalCtor):
+        fwdIter(*(&_index.fwd), MinimalCtor()),
+        bwdIter(*(&_index.rev), MinimalCtor())
+    {
+        fwdIter.setRevIter(bwdIter);
+        bwdIter.setRevIter(fwdIter);
+    }
+
+    template <typename TSpec2>
+    Iter(Iter<TBiIndex, VSTree<TopDown<ParentLinks<TSpec2> > > > const &_origin):
+        fwdIter(*(&_origin.fwdIter)),
+        bwdIter(*(&_origin.bwdIter))
+    {
+        fwdIter.setRevIter(bwdIter);
+        bwdIter.setRevIter(fwdIter);
+    }
+
+//____________________________________________________________________________
+
+    template <typename TSpec2>
+    inline Iter const &
+    operator = (Iter<TBiIndex, VSTree<TopDown<ParentLinks<TSpec2> > > > const &_origin)
+    {
+        fwdIter = *(&_origin.fwdIter);
+        bwdIter = *(&_origin.bwdIter);
+        return *this;
+    }
+};
+
+// Specialization of undirectional FM index iterator for use in the bidirectional FM index iterator
+template <typename TText, class TOccSpec, typename TLengthSum, typename TSpec>
+class Iter<Index<TText, FMIndex<TOccSpec, FMIndexConfig<TOccSpec, TLengthSum, FMBidirectional> > >, VSTree<TopDown<TSpec> > >
```

Comment:
The Iterator of a ```BidirectionalIndex<TIndex>``` (encapsulating a pair of ```TIndex```) should encapsulate a pair of Iterators of type ```TIndex```.